Clean up boot process

wycats opened the pull request:
This refactor improves the clarity and explicitness of the Application boot process. In particular, it takes
advantage of the previous ApplicationInstance refactoring, and explicitly exposes boot phases via promises.

The end-game here is to allow FastBoot to boot up an application but not create an application instance, and allow individual FastBoot requests to create instances on demand via the `visit()` API.

This will also help the testing infrastructure, which also wants to boot an application once for the entire test suite, and create an instance for each test run.

Commit:
Clean up boot process

This is the beginning of a refactor that improves the clarity and
explicitness of the Application boot process. In particular, it takes
advantage of the previous ApplicationInstance refactoring, and
explicitly exposes boot phases via promises.

The end-game here is to allow FastBoot to boot up an application but
not create an application instance, and allow individual FastBoot
requests to create instances on demand.

This will also help the testing infrastructure, which also wants to boot
an application once for the entire test suite, and create an instance
for each test run.

stefanpenner commented on packages/ember-application/lib/system/application.js:
```diff
@@ -256,28 +256,27 @@ var Application = Namespace.extend(DeferredMixin, {
   init: function() {
     this._super();
```
`this._super.apply(this, arguments);`

tomdale replied:
Hm. This code is not in a change in this commit.

stefanpenner commented on packages/ember-application/lib/system/application.js:
```diff
@@ -256,28 +256,27 @@ var Application = Namespace.extend(DeferredMixin, {
+      // tamper with the default `Ember.Router`.
+      // 2.0TODO: Can we move this into a globals-mode-only library?
+      this.Router = Router.extend();
+      this.waitForDOMReady(this.buildDefaultInstance());
```
i know several people who want to optionally begin before DOMReady, any thoughts on how to facilitate that officially?

chadhietala replied:
Would be great to not wait for DOM ready, especially when you [stream a chunked encoded](http://zoompf.com/blog/2012/05/too-chunky) index.html.  We optimize by streaming down JSON snippets to the index.html. Would be great to hide Ember boot underneath the initial connection.

tomdale replied:
Yes, this is absolutely one of the goals of the refactor: to give developers fine-grained control over the boot process. While I would prefer to preserve the current DOM ready semantics, our goal is that developers can prevent the default autobooting behavior, and use the same API as the FastBoot harness to control how the app is booted.

stefanpenner commented on packages/ember-application/lib/system/application.js:
```diff
@@ -671,20 +688,24 @@ var Application = Namespace.extend(DeferredMixin, {
+      }
 
-    this.resolve(this);
+      this.resolve(this);
```
is this bootResolve?

tomdale replied:
No, this is for the (deprecated) case of `Application.create().then(() => { })`. I don't think we can remove this until 2.0. Do you disagree?

Commit:
Legacy -> deprecated method naming

Shout out to @rwjblue

Commit:
Removes location test

This test had its heart in the right place, but the semantics it
actually exercises are unimportant and simply tested the side-effects
of a previous implementation.

To wit, this test sets the application’s location to `none`, yet still
expects the application’s `rootURL` to be injected on _any_ arbitrary
location looked up via the container.

Under the hood, the router registered a private (dash-prefixed)
injection, so we feel confident we can refactor the internal
implementation without breaking any apps.

In short, this test was testing the side-effects of a private injection,
which we are eliminating, while retaining the public API it supported.

Commit:
More granular control over routing start

Commit:
Implement Application’s `visit()` API

Commit:
Always resolve the boot promise, even in autoboot

dgeb committed:
Create a registry per ApplicationInstance

The instance's registry sets its fallback to the corresponding
application's registry.

dgeb committed:
Introduce fallback registry.

A registry's fallback registry resolves registrations and injections
when no matches can be found.

tomdale commented:
@rwjblue asked for a diagram of this change, thought it might be useful for others.
![untitled 3 2015-01-29 11-51-14](https://cloud.githubusercontent.com/assets/90888/5964988/747c8008-a7ad-11e4-861e-6f67e8c139fb.jpg)

dgeb committed:
Fix module name

dgeb committed:
Move validation of `fullName` parsing to `_parseName`.

Ensure that `name` and `type` are valid after initial parsing and before
any post-parsing mutations might be performed.

This fixes a problem in which mutations are performed to the cached
object (e.g. "route:basic" -> "route:") and then the validation of the
mutated object fail when it's accessed a second time.

It's also more efficient, since validation only occurs once.

Commit:
Use imported `run`

Commit:
Add feature flag for visit API

Commit:
Add test for disabling autoboot

Commit:
Adds simple visit() test

rwjblue commented on packages/ember-application/lib/system/application.js:
```diff
@@ -253,31 +254,44 @@ var Application = Namespace.extend(DeferredMixin, {
+    // decremented by the Application's own `initialize` method.
+    this._readinessDeferrals = 1;
+
+    if (!Ember.FEATURES.isEnabled('ember-application-visit') || this.autoboot) {
```
The feature flag system cannot handle this properly.

It **must** be precisely:

```javascript
if (Ember.FEATURES.isEnabled('ember-application-visit')) {
  /* stuff when enabled */
} else {
  /* stuff when disabled */
}
```

tomdale committed:
Fix feature flagged autoboot
